crates/uv-distribution-types/src/status_code_strategy.rs

```diff
@@ -0,0 +1,255 @@
+
+use crate::{IndexCapabilities, IndexUrl};
+
+#[derive(Debug, Clone, Default, Eq, PartialEq)]
+pub enum IndexStatusCodeStrategy {
+    #[default]
+    Default,
+    IgnoreErrorCodes {
+        status_codes: FxHashSet<StatusCode>,
+    },
+}
+
+impl IndexStatusCodeStrategy {
+    /// Derive a strategy from an index URL. We special-case PyTorch. Otherwise,
+    /// we follow the default strategy.
+    pub fn from_index_url(url: &Url) -> Self {
+        if url
+            .host_str()
+            .is_some_and(|host| host.ends_with("pytorch.org"))
+        {
+            // The PyTorch registry returns a 403 when a package is not found, so
+            // we ignore them when deciding whether to search other indexes.
+            Self::IgnoreErrorCodes {
+                status_codes: FxHashSet::from_iter([StatusCode::FORBIDDEN]),
+            }
+        } else {
+            Self::Default
+        }
+    }
+
+    /// Derive a strategy from a list of status codes to ignore.
+    pub fn from_ignored_error_codes(status_codes: &[SerializableStatusCode]) -> Self {
+        Self::IgnoreErrorCodes {
+            status_codes: status_codes
+                .iter()
+                .map(SerializableStatusCode::status_code)
+                .copied()
+                .collect::<FxHashSet<_>>(),
+        }
+    }
+
+    /// Based on the strategy, decide whether to continue searching the next index
+    /// based on the status code returned by this one.
+    pub fn handle_status_code(
+        &self,
+        status_code: StatusCode,
+        index_url: &IndexUrl,
+        capabilities: &IndexCapabilities,
+    ) -> IndexStatusCodeDecision {
+        capabilities.set_by_status_code(status_code, index_url);
+        match self {
+            IndexStatusCodeStrategy::Default => match status_code {
+                StatusCode::NOT_FOUND => IndexStatusCodeDecision::Continue,
+                StatusCode::UNAUTHORIZED => IndexStatusCodeDecision::Stop,
+                StatusCode::FORBIDDEN => IndexStatusCodeDecision::Stop,
+                _ => IndexStatusCodeDecision::Stop,
+            },
+            IndexStatusCodeStrategy::IgnoreErrorCodes { status_codes } => {
+                if status_codes.contains(&status_code) {
+                    IndexStatusCodeDecision::Continue
+                } else {
+                    IndexStatusCodeStrategy::Default.handle_status_code(
+                        status_code,
+                        index_url,
+                        capabilities,
+                    )
+                }
+            }
+        }
+    }
+}
+
+/// Decision on whether to continue searching the next index.
+#[derive(Debug, Clone, Copy, Hash, Eq, PartialEq)]
+pub enum IndexStatusCodeDecision {
+    Continue,
+    Stop,
+}
+
+#[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord, Hash)]
+pub struct SerializableStatusCode(StatusCode);
+
+impl SerializableStatusCode {
+    pub fn status_code(&self) -> &StatusCode {
+        &self.0
+    }
+}
```

Comment:
Should this just be a deref implementation? (we do that pretty often)

Reply:
I’m pretty sure you can’t match against e.g. StatusCode::FORBIDDEN that way (which is what we need this for)

Reply:
Wouldn't you just deref it where you're currently calling this method?

Reply:
Oh yeah, I was thinking of the transparent deref, but you're right you can just call `deref()`. Updated